File: 3_2_winform_protocol_structure_demo/3_2_winform_protocol_structure_demo/Protocol/WorkManager.cs
using System.Collections.Generic;

namespace HYKT
{
    class WorkManager
    {
        private List<Work> workList; 

        public WorkManager() {
            workList = new List<Work>();
        }

        public void AddWork(byte[] data)
        {
            Work work = new Work();
            work.Data = data;

            workList.Add(work);
        }

        public void CleanWorkList() {
            workList.Clear();
        }

        public void FinishAWork()
        {
            int count = workList.Count;
            if (count > 0)
            {
                workList.RemoveAt(0); //first work
            }
        }

        public Work GetCurrentWork() {
            int count = workList.Count;
            if (count > 0)
            {
                return workList[0]; //first work
            }
            return null;
        }

        public void markWorkFailed() {
            workList[0].FailedTimes += 1;

            if (workList[0].FailedTimes == 3) {
                //todo: call back for notice and abort all the work
            }
        }

        public Work GetWork() {
            Work work = null;
            //first in first out
            int count = workList.Count;
            if (count > 0) {
                work = workList[0];
            }

            return work;
        }

        public bool IsHasWork() {
            int count = workList.Count;

            if (count > 0) return true;
            else return false;
        }

    }
}

File: 13_3_stm32_iap_downloader_tool/src/protocol/workmanager.cpp


#include "workmanager.h"


WorkManager::WorkManager()
{


}

WorkManager::~WorkManager()
{

}

void WorkManager::AddWork(QByteArray &pkg) {
   Work *work = new Work();
   work->data = pkg;
   work_list_.append(work);
}

void WorkManager::InsertWorkAtHead(QByteArray &pkg) {
    Work *work = new Work();
    work->data = pkg;
    work_list_.insert(0, work);
}

void WorkManager::CleanWorkList() {
    for (int a = 0; a < work_list_.count(); ++a) {
        Work *work = work_list_.at(a);
        delete work;
    }
   work_list_.clear();
}

void WorkManager::FinishAWork()
{
   int count = work_list_.count();
   if (count > 0) {
       Work *work = work_list_.at(0);
       delete work;
       work_list_.removeAt(0); //first work
   }
}

Work* WorkManager::GetCurrentWork() {
   int count = work_list_.count();
   if (count > 0) {
       return work_list_.at(0); //first work
   }
   return nullptr;
}

bool WorkManager::IsHasWork() {
   int count = work_list_.count();

   if (count > 0) return true;
   else return false;
}

int WorkManager::GetWorkCount() {
    return work_list_.count();
}

